File: host/src/hidpin/_hidraw.py
"""Linux hidraw backend.

Mirrors the small part of the hidapi API that this library uses, but talks to
/dev/hidrawN directly. Unlike the libusb backend of the hidapi wheels it does
not have to detach the kernel driver, so it works with the plain udev rule and
leaves the device usable for other programs.
"""

from __future__ import annotations

import fcntl
import os
import select
from pathlib import Path

SYSFS_HIDRAW = Path("/sys/class/hidraw")
DEV_DIR = Path("/dev")

# Linux ioctl encoding (include/uapi/asm-generic/ioctl.h) for the HIDIOC* requests.
_IOC_WRITE = 1
_IOC_READ = 2
_HID_IOC_TYPE = ord("H")


def _ioc(direction: int, nr: int, size: int) -> int:
    return (direction << 30) | (size << 16) | (_HID_IOC_TYPE << 8) | nr


def _hidiocsfeature(size: int) -> int:
    return _ioc(_IOC_WRITE | _IOC_READ, 0x06, size)


def _hidiocgfeature(size: int) -> int:
    return _ioc(_IOC_WRITE | _IOC_READ, 0x07, size)


def _hidiocginput(size: int) -> int:
    return _ioc(_IOC_WRITE | _IOC_READ, 0x0A, size)


def _read_text(path: Path) -> str:
    try:
        return path.read_text().strip()
    except OSError:
        return ""


def _uevent_fields(path: Path) -> dict[str, str]:
    fields = {}
    for line in _read_text(path).splitlines():
        key, separator, value = line.partition("=")
        if separator:
            fields[key] = value
    return fields


def enumerate(vendor_id: int = 0, product_id: int = 0) -> list[dict]:
    """Lists hidraw devices, in the shape hid.enumerate() returns."""
    results = []
    for node in sorted(SYSFS_HIDRAW.glob("hidraw*")):
        fields = _uevent_fields(node / "device" / "uevent")
        hid_id = fields.get("HID_ID", "").split(":")
        if len(hid_id) != 3:
            continue
        try:
            vid, pid = int(hid_id[1], 16), int(hid_id[2], 16)
        except ValueError:
            continue
        if (vendor_id and vid != vendor_id) or (product_id and pid != product_id):
            continue

        usb_device = (node / "device").resolve().parent.parent
        results.append(
            {
                "path": str(DEV_DIR / node.name).encode(),
                "vendor_id": vid,
                "product_id": pid,
                "serial_number": fields.get("HID_UNIQ", "") or _read_text(usb_device / "serial"),
                "manufacturer_string": _read_text(usb_device / "manufacturer"),
                "product_string": _read_text(usb_device / "product") or fields.get("HID_NAME", ""),
                "usage_page": 0,  # the kernel does not expose it here; the caller checks the reports instead
                "usage": 0,
                "interface_number": -1,
            }
        )
    return results


class device:  # noqa: N801 - named after hid.device so it can stand in for it
    """One open hidraw device."""

    def __init__(self) -> None:
        self._fd: int | None = None

    def open_path(self, path) -> None:
        if isinstance(path, bytes):
            path = path.decode()
        self._fd = os.open(path, os.O_RDWR | os.O_NONBLOCK)

    def _require_fd(self) -> int:
        if self._fd is None:
            raise OSError("device is not open")
        return self._fd

    def read(self, length: int, timeout_ms: int = 0) -> list[int]:
        """Reads one report. timeout_ms < 0 waits forever, 0 returns immediately."""
        fd = self._require_fd()
        timeout = None if timeout_ms < 0 else timeout_ms / 1000.0
        ready, _, _ = select.select([fd], [], [], timeout)
        if not ready:
            return []
        return list(os.read(fd, length))

    def write(self, data) -> int:
        """Sends an output report; the first byte is the report ID."""
        return os.write(self._require_fd(), bytes(data))

    def send_feature_report(self, data) -> int:
        buffer = bytearray(data)
        fcntl.ioctl(self._require_fd(), _hidiocsfeature(len(buffer)), buffer)
        return len(buffer)

    def get_feature_report(self, report_id: int, length: int) -> list[int]:
        buffer = bytearray(length)
        buffer[0] = report_id
        count = fcntl.ioctl(self._require_fd(), _hidiocgfeature(length), buffer)
        return list(buffer[: count if count > 0 else 0])

    def get_input_report(self, report_id: int, length: int) -> list[int]:
        buffer = bytearray(length)
        buffer[0] = report_id
        count = fcntl.ioctl(self._require_fd(), _hidiocginput(length), buffer)
        return list(buffer[: count if count > 0 else 0])

    def close(self) -> None:
        if self._fd is not None:
            os.close(self._fd)
            self._fd = None


def available() -> bool:
    return SYSFS_HIDRAW.is_dir()

File: host-go/internal/hidraw/hidraw.go
//go:build linux

// Package hidraw talks to Linux /dev/hidrawN devices directly: enumeration through sysfs and
// reports through read, write and the HIDIOC* ioctls. It needs neither cgo nor hidapi, and unlike
// the libusb backend of hidapi it leaves the kernel driver attached.
package hidraw

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"syscall"
	"time"
	"unsafe"
)

// SysfsRoot and DevDir are variables so tests can point them at a fake tree.
var (
	SysfsRoot = "/sys/class/hidraw"
	DevDir    = "/dev"
)

// Info describes one hidraw device.
type Info struct {
	Path         string
	VendorID     uint16
	ProductID    uint16
	Serial       string
	Manufacturer string
	Product      string
}

func readText(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(data))
}

func ueventFields(path string) map[string]string {
	fields := map[string]string{}
	for _, line := range strings.Split(readText(path), "\n") {
		if key, value, ok := strings.Cut(line, "="); ok {
			fields[key] = value
		}
	}
	return fields
}

// Enumerate lists hidraw devices; a zero vendorID or productID matches any.
func Enumerate(vendorID, productID uint16) ([]Info, error) {
	entries, err := os.ReadDir(SysfsRoot)
	if err != nil {
		return nil, fmt.Errorf("cannot list %s: %w", SysfsRoot, err)
	}
	var infos []Info
	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasPrefix(name, "hidraw") {
			continue
		}
		device := filepath.Join(SysfsRoot, name, "device")
		fields := ueventFields(filepath.Join(device, "uevent"))
		ids := strings.Split(fields["HID_ID"], ":")
		if len(ids) != 3 {
			continue
		}
		vid, err1 := strconv.ParseUint(ids[1], 16, 32)
		pid, err2 := strconv.ParseUint(ids[2], 16, 32)
		if err1 != nil || err2 != nil {
			continue
		}
		if (vendorID != 0 && uint16(vid) != vendorID) || (productID != 0 && uint16(pid) != productID) {
			continue
		}

		// .../3-2/3-2:1.0/0003:1209:0001.000F -> the USB device is two levels up.
		usbDevice := ""
		if resolved, err := filepath.EvalSymlinks(device); err == nil {
			usbDevice = filepath.Dir(filepath.Dir(resolved))
		}
		serial := fields["HID_UNIQ"]
		if serial == "" && usbDevice != "" {
			serial = readText(filepath.Join(usbDevice, "serial"))
		}
		product := readText(filepath.Join(usbDevice, "product"))
		if product == "" {
			product = fields["HID_NAME"]
		}
		infos = append(infos, Info{
			Path:         filepath.Join(DevDir, name),
			VendorID:     uint16(vid),
			ProductID:    uint16(pid),
			Serial:       serial,
			Manufacturer: readText(filepath.Join(usbDevice, "manufacturer")),
			Product:      product,
		})
	}
	sort.Slice(infos, func(a, b int) bool { return infos[a].Path < infos[b].Path })
	return infos, nil
}

// Linux ioctl encoding (include/uapi/asm-generic/ioctl.h) for the HIDIOC* requests.
const (
	iocWrite = 1
	iocRead  = 2
)

func hidIOC(dir, nr, size uintptr) uintptr {
	return dir<<30 | size<<16 | uintptr('H')<<8 | nr
}

func hidiocSFeature(size int) uintptr { return hidIOC(iocWrite|iocRead, 0x06, uintptr(size)) }
func hidiocGFeature(size int) uintptr { return hidIOC(iocWrite|iocRead, 0x07, uintptr(size)) }
func hidiocGInput(size int) uintptr   { return hidIOC(iocWrite|iocRead, 0x0A, uintptr(size)) }

// Device is one open hidraw device.
type Device struct {
	file *os.File
}

// Open opens a hidraw device node such as /dev/hidraw0.
func Open(path string) (*Device, error) {
	file, err := os.OpenFile(path, os.O_RDWR, 0)
	if err != nil {
		return nil, err
	}
	return &Device{file: file}, nil
}

// Read reads one input report (report ID first). A negative timeout waits forever; it returns
// 0 and no error when the timeout expires.
func (d *Device) Read(buf []byte, timeout time.Duration) (int, error) {
	deadline := time.Time{}
	if timeout >= 0 {
		deadline = time.Now().Add(timeout)
	}
	if err := d.file.SetReadDeadline(deadline); err != nil {
		return 0, err
	}
	n, err := d.file.Read(buf)
	if errors.Is(err, os.ErrDeadlineExceeded) {
		return 0, nil
	}
	return n, err
}

// Write sends an output report; the first byte is the report ID.
func (d *Device) Write(report []byte) (int, error) {
	return d.file.Write(report)
}

func (d *Device) ioctl(request uintptr, buf []byte) (int, error) {
	conn, err := d.file.SyscallConn()
	if err != nil {
		return 0, err
	}
	var result uintptr
	var errno syscall.Errno
	err = conn.Control(func(fd uintptr) {
		result, _, errno = syscall.Syscall(syscall.SYS_IOCTL, fd, request, uintptr(unsafe.Pointer(&buf[0])))
	})
	if err != nil {
		return 0, err
	}
	if errno != 0 {
		return 0, errno
	}
	return int(result), nil
}

// GetFeatureReport reads a feature report; buf[0] selects the report ID and receives it back.
func (d *Device) GetFeatureReport(buf []byte) (int, error) {
	return d.ioctl(hidiocGFeature(len(buf)), buf)
}

// SendFeatureReport writes a feature report; the first byte is the report ID.
func (d *Device) SendFeatureReport(report []byte) (int, error) {
	buf := append([]byte(nil), report...)
	return d.ioctl(hidiocSFeature(len(buf)), buf)
}

// GetInputReport reads an input report on the control pipe; buf[0] selects the report ID.
func (d *Device) GetInputReport(buf []byte) (int, error) {
	return d.ioctl(hidiocGInput(len(buf)), buf)
}

// Close closes the device node.
func (d *Device) Close() error {
	return d.file.Close()
}
